org.eclipse.draw2d/src/org/eclipse/draw2d/EventListenerList.java

```diff
@@ -33,15 +39,7 @@ public synchronized <T> void addListener(Class<T> c, Object listener) {
 			throw new IllegalArgumentException();
 		}
 
-		int oldSize = (array == null) ? 0 : array.length;
-		Object[] newArray = new Object[oldSize + 2];
-		if (oldSize != 0) {
-			System.arraycopy(array, 0, newArray, 0, oldSize);
-		}
-		newArray[oldSize] = c;
-		oldSize++;
-		newArray[oldSize] = listener;
-		array = newArray;
+		listeners.computeIfAbsent(c, newC -> new ArrayList<>(2)).add(listener);
```

Comment:
That would probably have to be a thread-safe array, as indicated by the test failures. Though I wonder if this then removes the performance boost, as it now has to copy the internal array again...

```suggestion
		listeners.computeIfAbsent(c, newC -> new CopyOnWriteArrayList<>(2)).add(listener);
```

Reply:
Hmh. In the meantime I tried a version where I copy for the iterators. Which is a bit slower but modification save. The copy on write is interestingly faster until 10k of EditParts. Then my version of copying on access is faster. However copy on write is still faster then then the original version. So I guess we should use it or?

Reply:
Based on the snippet in the bug, I got the following results:

**Current Implementation**

2001 edit parts: 71 ms
10001 edit parts: 387 ms
20001 edit parts: 903 ms
40001 edit parts: 2955 ms
80001 edit parts: 9300 ms
160001 edit parts: 35599 ms

**ArrayList with Copy Constructor**

2001 edit parts: 66 ms
10001 edit parts: 198 ms
20001 edit parts: 365 ms
40001 edit parts: 1079 ms
80001 edit parts: 3001 ms
160001 edit parts: 9328 ms

**CopyOnWriteArrayList**

2001 edit parts: 72 ms
10001 edit parts: 242 ms
20001 edit parts: 559 ms
40001 edit parts: 1582 ms
80001 edit parts: 4893 ms
160001 edit parts: 16947 ms

So the initialization is the fastest when using a plan ArrayList, which is what one would expect. However, what the test doesn't cover is the access to those listeners, which should now be noticeably slower as a new array is created every time. So effectively, the original problem has been solved by moving it somewhere else.

To be honest, I' not a huge fan of either approaches and I wonder if there is an alternative we simply haven't considered yet...

Reply:
Hm... what about using a Deque/ConcurrentLinkedDeque instead of a List/(CopyOn)ArrayList?

We have free control over what data structure those listeners are stored in and performance-wise, it's very close to a plain ArrayList while also giving us thread-safety.

2001 edit parts: 73 ms
10001 edit parts: 240 ms
20001 edit parts: 435 ms
40001 edit parts: 1170 ms
80001 edit parts: 3629 ms
160001 edit parts: 10177 ms

Reply:
Hm... I'm not so fond of using linked lists. They have quite a memory overhead and the `EventListenerList` is used for every `EditPart` and `Figure`.

Reply:
What do you mean by "quite a memory overhead"? Based on the example with 160000 edit parts, I see the entire program taking ~600MB with the current implementation, 713MB when using an array list and 754MB when using a linked list, which to me is all in the same ballpark.
Regardless, if you think an array list is a better fit here, then that's fine by me.